dbt/adapters/spark/impl.py

```diff
@@ -133,44 +135,114 @@ def add_schema_to_cache(self, schema) -> str:
-                    f'Invalid value from "show table extended ...", '
-                    f"got {len(row)} values, expected 4"
+    def _get_relation_information_using_describe(self, row: agate.Row) -> Tuple[str, str, str]:
+        """Relation info fetched using SHOW TABLES and an auxiliary DESCRIBE statement"""
+        if len(row) == 3:
+            _schema, name, _ = row
+
+            table_name = f"{_schema}.{name}"
+            information = ""
+            try:
+                table_results = self.execute_macro(
+                    DESCRIBE_TABLE_EXTENDED_MACRO_NAME, kwargs={"table_name": table_name}
                 )
-            _schema, name, _, information = row
-            rel_type = RelationType.View if "Type: VIEW" in information else RelationType.Table
-            is_delta = "Provider: delta" in information
-            is_hudi = "Provider: hudi" in information
-            relation: BaseRelation = self.Relation.create(
+
+                for info_row in table_results:
+                    info_type, info_value, _ = info_row
+                    if not info_type.startswith("#"):
+                        information += f"{info_type}: {info_value}\n"
+            except dbt.exceptions.DbtRuntimeError as e:
+                logger.debug(f"Error while retrieving information about {table_name}: {e.msg}")
+
+            return _schema, name, information
+        else:
+            raise dbt.exceptions.DbtRuntimeError(
+                f'Invalid value from "show tables ...", got {len(row)} values, expected 3'
+            )
+
+    def _build_spark_relation_list(
+        self,
+        row_list: agate.Table,
+        relation_info_func: Callable[[agate.Row], Tuple[str, str, str]],
+    ) -> List[SparkRelation]:
+        """Aggregate relations with format metadata included."""
+        relations = []
+        for row in row_list:
+            _schema, name, information = relation_info_func(row)
+
+            rel_type: RelationType = (
+                RelationType.View if "Type: VIEW" in information else RelationType.Table
+            )
+            is_delta: bool = "Provider: delta" in information
+            is_hudi: bool = "Provider: hudi" in information
+            is_iceberg: bool = "Provider: iceberg" in information
+
+            relation: BaseRelation = self.Relation.create(  # type: ignore
                 schema=_schema,
                 identifier=name,
                 type=rel_type,
                 information=information,
                 is_delta=is_delta,
+                is_iceberg=is_iceberg,
                 is_hudi=is_hudi,
             )
             relations.append(relation)
 
         return relations
 
+    def list_relations_without_caching(
+        self, schema_relation: SparkRelation
+    ) -> List[SparkRelation]:
+        """Distinct Spark compute engines may not support the same SQL featureset. Thus, we must
+        try different methods to fetch relation information."""
+
+        try:
```

Comment:
All the refactoring I did was to make this function more traceable. How we at dbt Labs had let it grown over time meant any format addition or exception accounted for meant us veering into spaghetti code territory. Hard to read, hard to trust.

 With functions and comments, I do hope I've rendered this more approachable to the uninitiated. This will serve us the next time a format is to be added/adjusted.

Reply:
Hmmm, that's what caused this to be confusing in the first place imo. We cannot guarantee that 4-element scenarios flow from 3-element scenarios. We'll have to talk this out more together I think since I'm really hesitant to do this with my current perspective.

Reply:
Talked about this. 👍